### ISSUE TITLE
Generic initrd does not work with encrypted root FS without further configuration

### ISSUE THREAD

**Issue by LaszloGombos** [dracutdevs/dracut#2437]
Moving downstream discussion here upstream from https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1029324

with an encrypted root fs, the system will end up with a configuration which
looks something like this:

```
/etc/crypttab:
	<dev>_crypt UUID=<some_uuid> none luks,discard

/etc/fstab:
	/dev/mapper/<some_LVM_LV> / ext4 <options>

/boot/grub/grub.cfg:
	linux /vmlinuz-6.0.0-6-amd64 root=/dev/mapper/<some_LVM_LV> ro quiet
```

Now, after switching to dracut, the new initrd which is built will not
contain the necessary configuration to actually setup the "<dev>_crypt"
device, since autoconfiguration (rd.auto) is disabled and a hostonly
initrd (which would include a custom /etc/crypttab in the initrd) is not
built.

The result is an initrd which will hang waiting for
/dev/mapper/<some_LVM_LV> to show up, and which will eventually time
out.

A workaround for now is to change /etc/default/grub:
	GRUB_CMDLINE_LINUX_DEFAULT="quiet rd.auto=1"

This, however, means that the name of the crypt device will change from
"<dev>_crypt" (as specified in /etc/crypttab) to "luks-<some_uuid>",
which will result in ugly error messages during boot (that's mostly
cosmetic though) since systemd will fail to setup the device
"<dev>_crypt".

That, in turn, can be fixed by changing /etc/crypttab from:
	<dev>_crypt UUID=<some_uuid> none luks,discard
so that it reads:
	luks-<some_uuid> UUID=<some_uuid> none luks,discard

A further problem with building the generic initrd is that custom crypt
addons (like the fido2 support) won't be included in the initrd even if
fido2 is defined in /usr/lib/dracut/modules.d/90crypt/module-setup.sh
("fido2-device=auto").

CC @Alphix

**Comment by HNA-JP** [comment]
I'm not sure what I'm doing wrong, but with more-or-less the same setup I'm finding Dracut isn't respecting `/etc/crypttab.initramfs` or `/etc/crypttab.initrd` at all.

Dropping to the emergency shell shows `/etc/crypttab` in the initramfs image has something like:
```
luks-<LVM-LV-UUID>  /dev/dm-1    none  luks,discard
```

This means appending `rd.auto` to the GRUB cmdline works as mentioned as it'll attempt `rd.luks` with `rd.luks.crypttab=yes` by default.

I found including the `crypttab.initrd` I had by manually by renaming it to just `crypttab` and running `dracut -f --regenerate-all --include /etc/crypttab` didn't work either as the file either isn't included or gets overwritten part way through generation. 

Running dracut with `--print-cmdline` makes me think it would be a moot point even if I did manage to get the correct `crypttab` in the image however as the cmdline parameters bind to the LVM LV UUID and name `luks-<LVM-LV-UUID>` anyways.

My question is, is there something you need to do for dracut to respect `crypttab.initramfs` or `crypttab.initrd`?

**Comment by LaszloGombos** [comment]
@HNA-JP which version of dracut and which Linux distribution ? 

> Dracut isn't respecting /etc/crypttab.initramfs or /etc/crypttab.initrd

Can you please point to dracut documentation that would suggest to use crypttab.initramfs or crypttab.initrd ? Are you coming from another initrd generator where you used these before and now you expect that dracut works the same way ? 

If hostonly is what you want, you can still enable hostonly for your machine even if distribution defaults to non-hostonly. In hostonly mode dracut will respect `/etc/crypttab`.

The point of hostonly is that it is configured for your host. 
The point of generic is that it does not have knowledge of your host, so it needs additional configuration to work.

**Comment by HNA-JP** [comment]
> @HNA-JP which version of dracut and which Linux distribution ?

Ubuntu Mantic (23.10) with dracut-059-4ubuntu2

> Can you please point to dracut documentation that would suggest to use crypttab.initramfs or crypttab.initrd ? Are you coming from another initrd generator where you used these before and now you expect that dracut works the same way ?

Apologies- I seem to have misunderstood a couple things about dracut. I wasn't aware of what hostonly would do and assumed generic meant it wasn't distro specific.

I don't know what I did, but ironically I found dracut wasn't actually creating the necessary facilities to boot the system until I created a `crypttab.initramfs` file. When Googling around for dracut crypttab configs I had found a whole slew of threads to suggest it, but I'm now seeing there isn't any actual documentation on any of dracut's manpages, nor direct references to dracut in relation a lot of the times. I was misunderstanding posts about systemd-cryptsetup-generator, I think.

Running dracut with `hostonly` does indeed result in a working initramfs image however, so thank you for pointing me in the right direction.

On another note, generating an initramfs image with `hostonly` seems to result in duplicate `root` kernel parameters 
```
$ cat /proc/cmdline
BOOT_IMAGE=/vmlinuz-6.5.0-28-generic root=/dev/mapper/luks-<LVM-LV-UUID> ro init=/usr/lib/systemd/systemd rd.luks.name=<LVM-LV-UUID>=nvme0n1p5_crypt root=/dev/mapper/nvme0n1p5_crypt quiet splash vt.handoff=7
```

That's an issue for another thread, though.

Thanks again @LaszloGombos

**Comment by Alphix** [comment]
@LaszloGombos, unless I'm mistaken, the fix for this bug was commited (https://github.com/dracut-ng/dracut-ng/commit/2339acfaeee60d6bb26a1103db2e53bc8f9cb2d1) and subsequently reverted (https://github.com/dracut-ng/dracut-ng/commit/39eacd9223f2e8c14c76cbe885789a84df36a29b), so this bug should be reopened?

I'm certainly seeing the old behaviour on a Debian unstable test installation (i.e. it is necessary to boot with rd.auto=1, or the root-on-LVM-on-LUKS FS will never be mounted and the boot will hang)

**Comment by LaszloGombos** [comment]
@Alphix is this an issue only with non-hostonly ? Can you please try hostonly (without rd.auto=1) ?

**Comment by Alphix** [comment]
> [@Alphix](https://github.com/Alphix) is this an issue only with non-hostonly ? Can you please try hostonly (without rd.auto=1) ?

Yes, removing `rd.auto=1` from Grub's command line and regenerating the initrd with `hostonly` also works.

**Comment by adrelanos** [comment]
Follow-up ticket:
 
* https://github.com/dracut-ng/dracut-ng/issues/748

**Comment by camhux** [comment]
I just ran into this issue on Debian 13 after running `sudo apt install dracut` with no further intervention. Since my main partition is LUKS-encrypted, I was not able to boot until I figured out how to make the right edit in the GRUB boot screen. I'm a bit naïve to GNU/Linux as a personal computing platform, but this really exceeded the amount of risk I figured I was taking by installing a package from my distro's stable repository and was a super hostile first experience with `dracut`.
